### DISCUSSION TITLE
Exposure time calculator for pinhole cameras

### DISCUSSION THREAD

**Discussion by nylki** [can-lehmann/owlkettle#162] [Show and tell]
Here's the first thing I made with owlkettle :slightly_smiling_face: 

https://github.com/nylki/Pinhole-Exposure-Calculator

![PinholeCalculator Bildschirmfoto vom 2024-05-30 14-15-37](https://github.com/nylki/PinholeCalculator/assets/1710598/2c8d9f52-a543-4f97-a2c9-5459b0a90af6)

the app is intended to help with calculating exposure times for analog cameras and in particular pinhole cameras. It is probably not very useful for most people :D , but for me it was a fun little project.

My main motivation was both to learn more of GTK (and in particular adwaita) and a bit of nim as well. Owlkettle was fun to use, because of its declarative use, which reminded me a lot of SwiftUI or React (and Vue, etc.) that I have used heavily in the past.

**Comment by can-lehmann** [comment]
This is pretty cool, thanks for sharing! Did you hit any rough edges or missing features in owlkettle while building this?

**Reply by nylki** [reply]
thx, and you are welcome! In retrospect it went smooter than I expected and was in fact quite easy to get started with writing down the layout.

But there sure were a couple roadblocks and unclear things:

- For example I still don't quite understand why some things are controlled with nim Pragmas (eg. `{.expand: false.}`) instead of view props (like most other aspects).

- Also perhaps more GTK related (but not sure), I sometimes get stretched controls, like this switch. I would have expected that native controls keep their ideal aspect ratio. :smile: 

![stretched_controls](https://github.com/can-lehmann/owlkettle/assets/1710598/d0843439-4a51-41ca-97f6-6e9e7745f839)

**Reply by nylki** [reply]
And since you asked: one feature that would be really useful to have is a `Picture` widget that takes an image path as a prop and handles loading the image buffer by itself. Something like:

```.nim
Picture:
  path: "assets/someImage.png"
```

**Reply by can-lehmann** [reply]
> For example I still don't quite understand why some things are controlled with nim Pragmas (eg. {.expand: false.}) instead of view props (like most other aspects).

This seems to be a common point, probably because this feature doesn't really exist in web frameworks. Basically the difference is that widget fields set attributes on the widget itself, while properties (the pragmas) set attributes on the edge between the widget and its parent. So properties define how the widget is added to its parent. It might be worth it to write a detailed explanation for the docs.

> Also perhaps more GTK related (but not sure), I sometimes get stretched controls, like this switch. I would have expected that native controls keep their ideal aspect ratio. 😄

Yes, the behavior comes from GTK (although I am not sure if it is the default there). You can work round it quite easily though by using the `hAlign: AlignCenter` and `vAlign: AlignCenter` properties of the `Box.add` adder.
